### ISSUE TITLE
[BUG]: To send headers need dependency as Session Service

### ISSUE THREAD

**Issue by mustafa519** [phalcon/cphalcon#14959]
I've implemented my own DI container to use only need services.

I'm getting error when i'm trying to check security hashes on login page. I think, Phalcon triggers to call **Session Service** by itself when Phalcon try to send page headers like Cookies.

I don't want to use Session Service. Do I have to create session service for cookie and response services as dependency to send headers properly? Or am i missing something?

```php
// Type: Phalcon\Di\Exception
// Error: Service 'session' wasn't found in the dependency injection container
// File: /public/index.php
// Line 72:
// #0 [internal function]: Phalcon\Di->get('session', NULL)
// #1 [internal function]: Phalcon\Di->getShared('session')
// #2 [internal function]: Phalcon\Http\Cookie->send()
// #3 [internal function]: Phalcon\Http\Response\Cookies->send()
// #4 [internal function]: Phalcon\Http\Response->sendCookies()
// #5 /opt/lampp/htdocs/project/public/index.php(72): Phalcon\Mvc\Application->handle()
// #6 /opt/lampp/htdocs/project/public/index.php(120): Bootstrap->__construct()
// #7 {main}

echo $this->handle()->getContent(); # \Phalcon\Mvc\Application Line 72
```

## My DI Implement and Service Container

```php
$this->DI = new \Phalcon\Di();
$this->setDI($this->DI);
```

```php
$this->DI->setShared('config', $config);

$this->DI->setShared('db', function () use ($config)
{
  $CONNECTION = new \Phalcon\Db\Adapter\Pdo\Mysql(
    $config->APP->{$config->APP->STATUS}->MYSQL->toArray()
  );

  return $CONNECTION;
});

$this->DI->setShared('crypt', function() use ($config)
{
  $CRYPT = new \Phalcon\Crypt();
  $CRYPT->setCipher('aes-256-ctr');
  $CRYPT->setKey($config->APP->SECRET_KEY);

  return $CRYPT;
});

$this->DI->setShared('security', function()
{
  $SECURITY = new \Phalcon\Security();
  $SECURITY->setWorkFactor(10);

  return $SECURITY;
});

$this->DI->setShared('cookies', function()
{
  $COOKIES = new \Phalcon\Http\Response\Cookies();
  $COOKIES->useEncryption(true);

  return $COOKIES;

});

$this->DI->setShared('request', ['className' => 'Phalcon\Http\Request']);
$this->DI->setShared('response', ['className' => 'Phalcon\Http\Response']);

$this->DI->setShared('router', function ()
{
  return require APP_PATH .'/_config/Routes.php';
});
```

## My Login Page

```php
if ($this->security->checkHash($params->password, $data->USER_PASSWORD))
{
  $this->cookies->set(
    'auth',
    $data->USER_DATA,
    '/',
    true,
    $_SERVER['APP_STATUS'] === 'development'
      ? ''
      : ('.' .$_SERVER['HTTP_HOST']),
    true,
  );
}
```

## My Temporary Solution
```php
$this->DI->setShared('session', function()
{
    class Foo
    {
      public function __call($method, $args)
      {
        if (isset($this->$method))
        {
          $func = $this->$method;
          return call_user_func_array($func, $args);
      }
    }

    public function isstarted() { return true; }
  }

 $SESSION = new \Foo();

  return $SESSION;
});
```

**Details**
 - Phalcon version: 3.4.5
 - PHP Version: 7.3.14
 - Operating System: Ubuntu 18.04.4 LTS
 - Installation type: Compiling from source
 - Server: Apache

**Comment by jenovateurs** [comment]
Hi @mustafa519,
In the Phalcon code :  
https://github.com/phalcon/cphalcon/blob/d3e94630156a0c4441bf5f72e6ed071941c5908b/phalcon/Http/Cookie.zep#L473

We didn't check if the session service is defined.
So I think it's a bug to fix.
Thanks you for the detailed bug ;-).

**Comment by mustafa519** [comment]
Hello @jenovateurs. Thanks the quick reply.

I just really surprised that i'm the only one that who doesn't use Session Service around Phalcon developers. :smile: 

By the way, is there any trick for Cookie Service to set `SameSite` rule in `Phalcon 3.4.5` version?

**Comment by jenovateurs** [comment]
Hello @mustafa519, I saw the documentation v4.
And there is this note about session and cookie :
NOTE: If your DI container contains the session service, the cookies will be stored in the session automatically. If not, they will not be stored, and you are responsible to persist them if you wish to.

So maybe you can, but I didn't see how.
I just saw that the bug was already closed 15 days ago in phalcon:4.1.x
https://github.com/phalcon/cphalcon/issues/14944

About SameSite, I saw that it's possible with v4 : https://docs.phalcon.io/4.0/fr-fr/response#samesite
Can you try this trick with v3.4.5 ?

**Comment by mustafa519** [comment]
I've researched on Google and old Phalcon issues including closed. But i miss the issue, thank you for information.

According to `Cookie.zep` [v3.4.5](https://github.com/phalcon/cphalcon/blob/v3.4.5/phalcon/http/cookie.zep), there is no support to set options as array or something like that. So i think i should use native `setCookie` or to continue without `SameSite=Strict` rule for now.


There are many projects which prepared with `Phalcon v3.4.5`. That's why i can't upgrade Phalcon version now.

Thank you for helping me and for that great framework! Now issue can be closed.
